M2/Macaulay2/packages/SparseResultants.m2

```diff
@@ -462,6 +463,210 @@ genericMultihomogeneousPolynomial (VisibleList,VisibleList) := o -> (k,d) -> (
     sum(gens A,e -> (b = last baseName e; e * product(n,i -> (mm_i)_(b_i))))
 );
 
+-- Hyperdeterminants --
+
+MultidimensionalMatrix = new Type of HashTable;
+
+MultidimensionalMatrix#{Standard,AfterPrint} = MultidimensionalMatrix#{Standard,AfterNoPrint} = M -> (
+    << endl << concatenate(interpreterDepth:"o") << lineNumber << " : " << "Multidimensional matrix of format " << M#"format" << " over " << M#"ring" << endl;
+);
+
+net MultidimensionalMatrix := M -> net(M#"list");
+
+multidimensionalMatrix = method(TypicalValue => MultidimensionalMatrix);
+
+multidimensionalMatrix (List) := (L) -> (
+    X := makeRing L;
+    F := makeMultilinearForm(L,X);
+    L' := makeList(F,X);
+    new MultidimensionalMatrix from {
+        "format" => apply(X,i -> #i),
+        "list" => L',
+        "multilinearForm" => F,
+        "varsMultilinearForm" => X,
+        "ring" => coefficientRing ring F  
+    }
+);
+
+Thing * MultidimensionalMatrix := (a,M) -> (
+    applyToEntries(M,m -> a*m)
+);
+
+RingElement * MultidimensionalMatrix := (a,M) -> (
+    applyToEntries(M,m -> a*m)
+);
+
+- MultidimensionalMatrix := (M) -> (-1) * M;
+
+MultidimensionalMatrix - MultidimensionalMatrix := (M,N) -> (
+   if M#"format" != N#"format" then error "expected matrix of the same format";
+   multidimensionalMatrix(M#"list" - N#"list")
+);
+
++ MultidimensionalMatrix := (M) -> M;
+
+MultidimensionalMatrix + MultidimensionalMatrix := (M,N) -> (
+   if M#"format" != N#"format" then error "expected matrix of the same format";
+   multidimensionalMatrix(M#"list" + N#"list")
+);
+
+MultidimensionalMatrix == MultidimensionalMatrix := (M,N) -> (
+   if M#"format" != N#"format" then return false;
+   if M#"ring" =!= N#"ring" then return false;
+   M#"multilinearForm" == N#"multilinearForm"
+);
+
+transposition = method();
+transposition (MultidimensionalMatrix,List) := (M,l) -> (
+    n := M#"format";
+    try assert(ring matrix {l} === ZZ and sort l == toList(0 .. #n-1)) else error("expected a permutation of the set "|toString toList(0 .. #n-1));
```

Comment:
It's better to use `if not ... then ... ` here instead of `try assert ... else ...`, because it's more logical, and because other errors will be masked.  For example, what if n is not a list, so #n gives an error?  Then you'll never see it.

Reply:
Now I left only the following `try assert(ring matrix {l} === ZZ) else error ... ` It should fail exactly when l is not a list of integers.

Reply:
The purpose of that line will be unclear to the reader.  Just use something like
```m2
if not all(l, i -> instance(i,ZZ)) then error ...
```

Reply:
Ok, changed.